File: mobile-client/apps/frutsmart-plant/legacy-disabled/skybolt-local-disabled/android/src/main/java/expo/modules/skybolt/SkyboltModule.kt
package expo.modules.skybolt

import androidx.core.net.toUri
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.skybolt.azureblob.crypto.FastMD5
import expo.modules.skybolt.core.bridge.ErrorMapper
import expo.modules.skybolt.core.config.CloudUploadSettings
import expo.modules.skybolt.core.events.Events
import expo.modules.skybolt.core.events.NativeEventSink
import expo.modules.skybolt.core.facade.SkyboltManager
import expo.modules.skybolt.core.upload.api.SessionConfig
import expo.modules.skybolt.core.util.logger
import expo.modules.skybolt.core.util.LogLevel
import expo.modules.skybolt.core.util.AppLogger
import expo.modules.skybolt.core.auth.AuthEnvironment
import expo.modules.skybolt.core.auth.toAuthTokens
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Skybolt Native Module - Cloud upload functionality for React Native.
 * 
 * Provides comprehensive file upload capabilities with:
 * - Chunked uploads to Azure Blob Storage
 * - Background upload with WorkManager
 * - Progress tracking and event emission
 * - Automatic retry with exponential backoff
 * - SAS token management
 */
class SkyboltModule : Module() {
    private val log by logger()
    
    /**
     * Module definition with all exposed functions and events.
     */
    override fun definition() = ModuleDefinition {
        Name("Skybolt")
        
        // ====================================================================
        // Events
        // ====================================================================
        
        Events("onUploadEvent")
        
        // ====================================================================
        // Lifecycle
        // ====================================================================
        
        OnCreate {
            val ctx = appContext.reactContext?.applicationContext
                ?: error("Application context is null")

            val prefix = "Skybolt"
            val isDebug = BuildConfig.DEBUG
            val application = ctx as? android.app.Application
            val minLevel = if (isDebug) LogLevel.VERBOSE else LogLevel.WARN

            AppLogger.init(
                prefix = prefix,
                isDebug = isDebug,
                minLevel = minLevel,
                application = application
            )

            SkyboltManager.initialize(ctx)

            Events.setSink(NativeEventSink { type, payload ->
                val eventData = buildMap<String, Any?> {
                    put("type", type)
                    putAll(payload)
                }
                sendEvent("onUploadEvent", eventData)
            })

            log.i { "Skybolt module created (event sink configured)" }
        }
        
        OnDestroy {
            // Clear event sink to prevent stale references
            Events.clear()
            
            // Cleanup development resources (Fast Refresh support)
            // This does NOT cancel uploads, only cleans up listeners
            SkyboltManager.cleanupForDevelopment()
            
            log.i { "Skybolt module destroyed (uploads continue in background)" }
        }
        
        // ====================================================================
        // Configuration
        // ====================================================================
        
        /**
         * Configure Skybolt with upload settings.
         * Must be called before starting any uploads.
         * 
         * @param settingsMap Configuration object from JavaScript
         */
        AsyncFunction("configure") Coroutine { settingsMap: Map<String, Any?> ->
            try {
                log.d { "configure() called with settings" }
                
                val settings = CloudUploadSettings.fromMap(settingsMap)
                SkyboltManager.configure(settings)
                
                log.i { "Skybolt configured successfully" }
            } catch (e: Exception) {
                log.e(e) { "Configuration failed" }
                throw IllegalArgumentException("Configuration failed: ${e.message}", e)
            }
        }
        
        // ====================================================================
        // Session Management
        // ====================================================================
        
        /**
         * Initialize a new upload session.
         * Creates session state and prepares for upload.
         * 
         * @param configMap Session configuration from JavaScript
         */
        AsyncFunction("initializeSession") Coroutine { configMap: Map<String, Any?> ->
            try {
                log.d { "initializeSession() called" }
                
                val config = SessionConfig.fromJsMap(configMap)
                SkyboltManager.initializeSession(config)
                
                log.i { "Session initialized: ${config.sessionId}" }
            } catch (e: Exception) {
                log.e(e) { "Session initialization failed" }
                throw IllegalArgumentException("Session initialization failed: ${e.message}", e)
            }
        }
        
        /**
         * Start an initialized upload session.
         * Begins uploading files in background.
         * 
         * @param sessionId Unique session identifier
         */
        AsyncFunction("startSession") Coroutine { sessionId: String ->
            try {
                require(sessionId.isNotBlank()) { "sessionId cannot be blank" }
                log.d { "startSession() called: $sessionId" }
                
                SkyboltManager.startSession(sessionId)
                
                log.i { "Session started: $sessionId" }
            } catch (e: Exception) {
                log.e(e) { "Failed to start session: $sessionId" }
                throw IllegalArgumentException("Failed to start session: ${e.message}", e)
            }
        }
        
        /**
         * Pause an active upload session.
         * 
         * @param sessionId Unique session identifier
         */
        AsyncFunction("pauseSession") Coroutine { sessionId: String ->
            try {
                require(sessionId.isNotBlank()) { "sessionId cannot be blank" }
                log.d { "pauseSession() called: $sessionId" }
                
                SkyboltManager.pauseSession(sessionId)
                
                log.i { "Session paused: $sessionId" }
            } catch (e: Exception) {
                log.e(e) { "Failed to pause session: $sessionId" }
                throw IllegalArgumentException("Failed to pause session: ${e.message}", e)
            }
        }
        
        /**
         * Resume a paused upload session.
         * 
         * @param sessionId Unique session identifier
         */
        AsyncFunction("resumeSession") Coroutine { sessionId: String ->
            try {
                require(sessionId.isNotBlank()) { "sessionId cannot be blank" }
                log.d { "resumeSession() called: $sessionId" }
                
                SkyboltManager.resumeSession(sessionId)
                
                log.i { "Session resumed: $sessionId" }
            } catch (e: Exception) {
                log.e(e) { "Failed to resume session: $sessionId" }
                throw IllegalArgumentException("Failed to resume session: ${e.message}", e)
            }
        }
        
        /**
         * Cancel an upload session.
         * Stops upload and cleans up resources.
         * 
         * @param sessionId Unique session identifier
         */
        AsyncFunction("cancelSession") Coroutine { sessionId: String ->
            try {
                require(sessionId.isNotBlank()) { "sessionId cannot be blank" }
                log.d { "cancelSession() called: $sessionId" }
                
                SkyboltManager.cancelSession(sessionId)
                
                log.i { "Session canceled: $sessionId" }
            } catch (e: Exception) {
                log.e(e) { "Failed to cancel session: $sessionId" }
                throw IllegalArgumentException("Failed to cancel session: ${e.message}", e)
            }
        }
        
        // ====================================================================
        // Session Queries
        // ====================================================================
        
        /**
         * Get current progress for a session.
         * Returns null if session doesn't exist.
         * 
         * @param sessionId Unique session identifier
         * @return Progress map or null
         */
        AsyncFunction("getSessionProgress") Coroutine { sessionId: String ->
            try {
                require(sessionId.isNotBlank()) { "sessionId cannot be blank" }
                log.d { "getSessionProgress() called: $sessionId" }
                
                val progress = SkyboltManager.getSessionProgress(sessionId)
                progress?.toJsMap()
            } catch (e: Exception) {
                log.e(e) { "Failed to get session progress: $sessionId" }
                null
            }
        }
        
        /**
         * List all active (non-terminal) session IDs.
         * 
         * @return List of session IDs
         */
        AsyncFunction("listActiveSessions") Coroutine { ->
            try {
                log.d { "listActiveSessions() called" }
                
                SkyboltManager.listActiveSessions()
            } catch (e: Exception) {
                log.e(e) { "Failed to list active sessions" }
                emptyList<String>()
            }
        }
        
        /**
         * List all pending sessions (PAUSED or PREPARING state).
         * Used to show recoverable uploads in UI after app restart.
         * 
         * @return List of session details
         */
        AsyncFunction("listPendingSessions") Coroutine { ->
            try {
                log.d { "listPendingSessions() called" }
                
                SkyboltManager.listPendingSessions()
            } catch (e: Exception) {
                log.e(e) { "Failed to list pending sessions" }
                emptyList<Map<String, Any>>()
            }
        }
        
        /**
         * Notify that authentication has been refreshed.
         * Triggers auto-resume for sessions paused due to auth expiration.
         */
        AsyncFunction("notifyAuthRefreshed") Coroutine { ->
            try {
                log.d { "notifyAuthRefreshed() called" }
                
                SkyboltManager.notifyAuthRefreshed()
                
                log.i { "Auth refresh notification processed" }
            } catch (e: Exception) {
                log.e(e) { "Failed to notify auth refresh" }
                throw e
            }
        }
        
        /**
         * Resume all pending sessions (PAUSED or PREPARING state).
         * Useful for batch resuming after app restart or user action.
         * 
         * @return Number of sessions resumed
         */
        AsyncFunction("resumeAllPending") Coroutine { ->
            try {
                log.d { "resumeAllPending() called" }
                
                val count = SkyboltManager.resumeAllPending()
                
                log.i { "Resumed $count pending sessions" }
                count
            } catch (e: Exception) {
                log.e(e) { "Failed to resume all pending sessions" }
                0
            }
        }
        
        // ====================================================================
        // Maintenance
        // ====================================================================
        
        /**
         * Purge completed sessions older than specified time.
         * 
         * @param olderThanMs Time threshold in milliseconds (default: 0 = all)
         * @return Number of sessions purged
         */
        AsyncFunction("purgeCompletedSessions") Coroutine { olderThanMs: Double? ->
            try {
                val threshold = olderThanMs?.toLong() ?: 0L
                log.d { "purgeCompletedSessions() called: threshold=${threshold}ms" }
                
                SkyboltManager.purgeCompletedSessions(threshold)
            } catch (e: Exception) {
                log.e(e) { "Failed to purge sessions" }
                0
            }
        }
        
        /**
         * Clean up temporary files.
         * 
         * @return Number of files cleaned
         */
        AsyncFunction("cleanupTempFiles") Coroutine { ->
            try {
                log.d { "cleanupTempFiles() called" }
                
                SkyboltManager.cleanupTempFiles()
            } catch (e: Exception) {
                log.e(e) { "Failed to cleanup temp files" }
                0
            }
        }

        // ====================================================================
        // Authentication
        // ====================================================================

        /**
          * Update authentication tokens from JavaScript.
          * 
          * @param map Auth tokens map
        */
        AsyncFunction("setAuthTokens") Coroutine { map: Map<String, Any?> ->
            val tokens = map.toAuthTokens()
            AuthEnvironment.manager.updateTokens(tokens)
            log.i { "Auth tokens updated from JS" }
        }

        /**
          * Get valid access token, refreshing if needed.
          * Returns null if no valid token is available.
        */
        AsyncFunction("getValidAccessToken") Coroutine { ->
            withContext(Dispatchers.IO) {
                AuthEnvironment.manager.getValidAccessTokenOrNull()
            }
        }

        /**
          * Clear stored authentication tokens.
        */
        AsyncFunction("clearAuthTokens") Coroutine { ->
            AuthEnvironment.manager.clear()
            log.i { "Auth tokens cleared" }
        }

        // ====================================================================
        // Utilities
        // ====================================================================
        
        /**
         * Extract MD5 hashes from local files.
         * 
         * @param fileUris Array of local file URIs
         * @return Array of MD5 result objects
         */
        AsyncFunction("extractMD5FromFiles") Coroutine { fileUris: List<String> ->
            try {
                require(fileUris.isNotEmpty()) { "fileUris cannot be empty" }
                log.d { "extractMD5FromFiles() called: count=${fileUris.size}" }
                
                val results = FastMD5.computeMd5HexBatch(
                    ctx = appContext.reactContext!!,
                    uris = fileUris
                )
                
                results.map { result ->
                    mapOf(
                        "uri" to result.uri,
                        "md5Hex" to result.md5Hex,
                        "sizeBytes" to result.sizeBytes,
                        "contentType" to result.contentType,
                        "lastModifiedMs" to result.lastModifiedMs
                    )
                }
            } catch (e: Exception) {
                log.e(e) { "Failed to extract MD5 hashes" }
                throw IllegalArgumentException("Failed to extract MD5: ${e.message}", e)
            }
        }
    }
}

File: mobile-client/packages/skybolt/src/skybolt/SkyboltModule.ts
import {NativeEventEmitter} from 'react-native';
import type {EventSubscription} from 'react-native';
import NativeSkybolt from '../NativeSkybolt';
import type {
  KeyValuePair,
  NativeCloudUploadSettings,
  NativeSessionConfig,
  NativeUploadEvent as NativeUploadEventSpec,
} from '../NativeSkybolt';
import type {
  AuthTokens,
  CloudUploadSettings,
  ItemProgress,
  Md5HexResult,
  NativeUploadEvent,
  PauseReason,
  PendingSession,
  SessionConfig,
  SessionProgress,
  UploadError,
  UploadEvent,
  UploadEventListener,
} from './Skybolt.types';

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Native module interface for Skybolt upload functionality.
 * Uses Expo Modules API (avoids deprecated NativeModule pattern).
 */

// ============================================================================
// Module Initialization
// ============================================================================

type SkyboltModule = NonNullable<typeof NativeSkybolt>;

function toPairs(record?: Record<string, string>): KeyValuePair[] | undefined {
  if (!record) {
    return undefined;
  }
  return Object.entries(record).map(([key, value]) => ({key, value}));
}

function toNativeSettings(settings: CloudUploadSettings): NativeCloudUploadSettings {
  return {
    ...settings,
    backend: {
      ...settings.backend,
      defaultHeaders: toPairs(settings.backend.defaultHeaders),
    },
  };
}

function toNativeSessionConfig(config: SessionConfig): NativeSessionConfig {
  return {
    ...config,
    items: config.items.map(item => ({
      ...item,
      metadata: toPairs(item.metadata),
    })),
  };
}

function toNativeAuthTokens(tokens: AuthTokens): AuthTokens {
  return {
    ...tokens,
    accessExpiresAtMs: tokens.accessExpiresAtMs,
    refreshExpiresAtMs: tokens.refreshExpiresAtMs,
  };
}

function toNativeEvent(native: NativeUploadEventSpec): NativeUploadEvent {
  return native as NativeUploadEvent;
}

// Configuration
type SkyboltApi = {
  // Configuration
  configure(settings: NativeCloudUploadSettings): Promise<void>;

  // Session management
  initializeSession(config: NativeSessionConfig): Promise<void>;
  startSession(sessionId: string): Promise<void>;
  pauseSession(sessionId: string): Promise<void>;
  resumeSession(sessionId: string): Promise<void>;
  cancelSession(sessionId: string): Promise<void>;

  // Session queries
  getSessionProgress(sessionId: string): Promise<SessionProgress | null>;
  listActiveSessions(): Promise<string[]>;
  listPendingSessions(): Promise<PendingSession[]>;
  resumeAllPending(): Promise<number>;

  // Auth management
  notifyAuthRefreshed(): Promise<void>;
  setAuthTokens(tokens: AuthTokens): Promise<void>;
  getValidAccessToken(): Promise<string | null>;
  clearAuthTokens(): Promise<void>;

  // Maintenance
  purgeCompletedSessions(olderThanMs?: number): Promise<number>;
  cleanupTempFiles(): Promise<number>;

  // Utilities
  extractMD5FromFiles(fileUris: string[]): Promise<Md5HexResult[]>;
  addListener(eventType: string): void;
  removeListeners(count: number): void;
};

let skyboltModule: SkyboltModule | null;

try {
  skyboltModule = NativeSkybolt;
} catch {
  skyboltModule = null;
}

/**
 * Check if native module is available.
 * Use this before calling module functions to handle graceful degradation.
 */
export const isAvailable = !!skyboltModule;

/**
 * Assert that native module is loaded.
 * @throws Error if module is not available
 */
function assertModule(): SkyboltApi {
  if (!skyboltModule) {
    throw new Error(
      '[Skybolt] Native module not found. Did you build the app?'
    );
  }
  return skyboltModule as unknown as SkyboltApi;
}

// ============================================================================
// Event Conversion
// ============================================================================

/**
 * Ensure required field exists in native event.
 * @throws Error if value is null or undefined
 */
const mustExist = <T>(v: T | null | undefined, label: string): T => {
  if (v == null) {
    throw new Error(`[Skybolt] Missing ${label} in native event`);
  }
  return v;
};

/**
 * Convert native upload event to standardized UploadEvent format.
 * Handles all event types and normalizes error information.
 */
export function toUploadEvent(native: NativeUploadEvent): UploadEvent {
  const eventType = native.type;

  switch (eventType) {
    // Simple session events (sin payload extra)
    case 'session:started':
    case 'session:resumed':
    case 'session:completed':
    case 'session:canceled':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
      };

    // Session paused with reason
    case 'session:paused':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        reason: (native.reason ?? 'user') as PauseReason,
      };

    // Session failure
    case 'session:failed':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        error: {
          code: native.errorCode ?? 'E_UNKNOWN',
          message: native.errorMessage ?? 'Unknown error',
        },
      };

    // Item completion
    case 'item:completed':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          clientItemId: mustExist(native.clientItemId, 'clientItemId'),
        },
      };

    // Item failure
    case 'item:failed':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          clientItemId: mustExist(native.clientItemId, 'clientItemId'),
          error: {
            code: native.errorCode ?? 'E_UNKNOWN',
            message: native.errorMessage ?? 'Unknown error',
          },
        },
      };

    // Auth required (sessionId puede venir null)
    case 'auth:required':
      return {
        type: eventType,
        sessionId: native.sessionId ?? undefined,
        pendingSessions: native.pendingSessions ?? [],
      };

    // Specific error events (sin retry extra)
    case 'error:forbidden':
    case 'error:contract':
    case 'error:checksum':
    case 'error:file-access':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          clientItemId: mustExist(native.clientItemId, 'clientItemId'),
          message: native.errorMessage ?? native.message ?? 'Error occurred',
        },
      };

    // Throttling / rate limiting
    case 'error:rate-limited':
    case 'error:throttled':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          clientItemId: mustExist(native.clientItemId, 'clientItemId'),
          message: native.errorMessage ?? native.message ?? 'Throttled',
          retryAfterMs: native.retryAfterMs ?? 0,
        },
      };

    // Network error
    case 'error:network':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          clientItemId: mustExist(native.clientItemId, 'clientItemId'),
          message: native.errorMessage ?? native.message ?? 'Network error',
          attempt: native.attempt ?? 0,
        },
      };

    // Fatal error (error:fatal)
    case 'error:fatal':
      return {
        type: 'error:fatal',
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          message: native.errorMessage ?? native.message ?? 'Fatal error',
          // necesitas stack?: string en NativeUploadEvent para esto
          stack: native.stack,
        },
      };

    // Item progress with detailed metrics
    case 'item:progress': {
      const payload: ItemProgress = {
        sessionId: mustExist(native.sessionId, 'sessionId'),
        clientItemId: mustExist(native.clientItemId, 'clientItemId'),
        bytesUploaded: native.bytesUploaded ?? 0,
        totalBytes: native.totalBytes ?? 0,
        blockIndex: native.blockIndex,
        blockSize: native.blockSize,
        retries: native.retries,
      };
      return {
        type: eventType,
        sessionId: payload.sessionId,
        payload,
      };
    }

    // SAS token events
    case 'sas:requested':
    case 'sas:received':
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        payload: {
          blobNames: native.blobNames ?? [],
        },
      };

    // SAS error
    case 'sas:error': {
      const error: UploadError = {
        code: native.errorCode ?? 'E_SAS',
        message: native.errorMessage ?? 'SAS error',
      };
      return {
        type: eventType,
        sessionId: mustExist(native.sessionId, 'sessionId'),
        error,
      };
    }

    // Upload state change (app restart, etc.)
    case 'upload:state-change':
      return {
        type: 'upload:state-change',
        sessionId: mustExist(native.sessionId, 'sessionId'),
        newState: mustExist(native.newState, 'newState'),
        oldState: mustExist(native.oldState, 'oldState'),
        // reason puede ser algo como "app_restart" o "network"
        reason: native.reason,
      };

    // Recovery complete
    case 'upload:recovery-complete':
      return {
        type: 'upload:recovery-complete',
        payload: {
          totalScanned: native.totalScanned ?? 0,
          pendingCount: native.pendingCount ?? 0,
        },
      };

    // Resume-all complete
    case 'upload:resume-all-complete':
      return {
        type: 'upload:resume-all-complete',
        payload: {
          totalPending: native.totalPending ?? 0,
          resumed: native.resumed ?? 0,
          failed: native.failed ?? 0,
        },
      };

    // Debug messages
    case 'debug':
      return {
        type: eventType,
        sessionId: native.sessionId ?? undefined,
        message: native.message ?? '',
      };

    // Unknown event types (dev error handling)
    default: {
      if (__DEV__) {
        throw new Error(
          `[Skybolt] Unknown native event type: ${String(eventType)}`
        );
      }
      return {
        type: 'debug',
        sessionId: native.sessionId ?? undefined,
        message: `Unknown event: ${String(eventType)}`,
      };
    }
  }
}

// ============================================================================
// Event Subscription API
// ============================================================================

export type { EventSubscription as Subscription };

/**
 * Subscribe to upload events from native module.
 * @param listener - Callback to handle upload events
 * @returns Subscription object with remove() method
 * 
 * @example
 * const subscription = addUploadListener((event) => {
 *   console.log('Upload event:', event);
 * });
 * // Later...
 * subscription.remove();
 */
export function addUploadListener(
  listener: UploadEventListener
): EventSubscription {
  const module = assertModule();
  const emitter = new NativeEventEmitter(module as never);

  return emitter.addListener('onUploadEvent', (nativeEvent: NativeUploadEventSpec) => {
    const native = toNativeEvent(nativeEvent);
    try {
      listener(toUploadEvent(native));
    } catch (error) {
      console.error('[Skybolt] Event conversion error:', error, native);
    }
  }) as EventSubscription;
}

// ============================================================================
// Configuration Utilities
// ============================================================================

/**
 * Apply default values to upload settings.
 * Validates required fields and sets sensible defaults.
 */
export async function applyDefaults(
  settings: CloudUploadSettings
): Promise<CloudUploadSettings> {
  const config = { ...settings };

  // Azure defaults
  config.azure ??= {
    serviceVersion: '2023-11-03',
    sendBlockMd5: true,
    defaultChunkBytes: 4 * 1024 * 1024, // 4MB
  };

  // Ensure minimum chunk size (64KB)
  if (config.azure.defaultChunkBytes < 64 * 1024) {
    config.azure.defaultChunkBytes = 64 * 1024;
  }

  // Concurrency defaults
  config.concurrency ??= {};
  config.concurrency.maxParallelFiles ??= 2;
  config.concurrency.maxParallelChunks ??= 4;

  // Retry defaults
  config.retry ??= {};
  config.retry.maxRetries ??= 3;
  config.retry.baseDelayMs ??= 500;
  config.retry.maxDelayMs ??= 10_000;

  // Validate required fields
  if (!config.backend?.baseUrl) {
    throw new Error('[Skybolt] backend.baseUrl is required');
  }

  const existBackendEndpoints =
    config.backend.endpoints?.sasBatchPath &&
    config.backend.endpoints?.sasRefreshPath;

  if (!existBackendEndpoints) {
    throw new Error(
      '[Skybolt] backend.endpoints.sasBatchPath and sasRefreshPath are required'
    );
  }

  return config;
}

// ============================================================================
// Configuration API
// ============================================================================

/**
 * Configure Skybolt with upload settings.
 * Must be called before starting any uploads.
 */
export async function configure(settings: CloudUploadSettings): Promise<void> {
  const configWithDefaults = await applyDefaults(settings);
  return assertModule().configure(toNativeSettings(configWithDefaults));
}

// ============================================================================
// Session Management API
// ============================================================================

/**
 * Initialize a new upload session with specified configuration.
 */
export async function initializeSession(config: SessionConfig): Promise<void> {
  return assertModule().initializeSession(toNativeSessionConfig(config));
}

/**
 * Start an initialized upload session.
 */
export async function startSession(sessionId: string): Promise<void> {
  if (!sessionId) {
    throw new Error('[Skybolt] sessionId required');
  }
  return assertModule().startSession(sessionId);
}

/**
 * Pause an active upload session.
 */
export async function pauseSession(sessionId: string): Promise<void> {
  if (!sessionId) {
    throw new Error('[Skybolt] sessionId required');
  }
  return assertModule().pauseSession(sessionId);
}

/**
 * Resume a paused upload session.
 */
export async function resumeSession(sessionId: string): Promise<void> {
  if (!sessionId) {
    throw new Error('[Skybolt] sessionId required');
  }
  return assertModule().resumeSession(sessionId);
}

/**
 * Cancel an active upload session.
 * This will stop all uploads and clean up resources.
 */
export async function cancelSession(sessionId: string): Promise<void> {
  if (!sessionId) {
    throw new Error('[Skybolt] sessionId required');
  }
  return assertModule().cancelSession(sessionId);
}

// ============================================================================
// Session Query API
// ============================================================================

/**
 * Get current progress information for a session.
 * @returns SessionProgress or null if session not found
 */
export async function getSessionProgress(
  sessionId: string
): Promise<SessionProgress | null> {
  if (!sessionId) {
    throw new Error('[Skybolt] sessionId required');
  }
  return assertModule().getSessionProgress(sessionId);
}

/**
 * List all currently active session IDs.
 */
export async function listActiveSessions(): Promise<string[]> {
  return assertModule().listActiveSessions();
}

/**
 * List all pending sessions (PAUSED or PREPARING state).
 * Useful for showing recoverable uploads in UI after app restart.
 * @returns Array of session details with progress information
 */
export async function listPendingSessions(): Promise<Array<{
  sessionId: string;
  status: string;
  uploadedBytes: number;
  totalBytes: number;
  itemCount: number;
  startedAt: number;
  endedAt: number;
}>> {
  return assertModule().listPendingSessions();
}

/**
 * Resume all pending sessions (PAUSED or PREPARING state).
 * Useful for batch resuming after app restart or user action.
 * @returns Number of sessions resumed
 */
export async function resumeAllPending(): Promise<number> {
  return assertModule().resumeAllPending();
}

// ============================================================================
// Auth Management API
// ============================================================================

/**
 * Notify native module that authentication has been refreshed.
 * Triggers auto-resume for sessions paused due to auth expiration.
 */
export async function notifyAuthRefreshed(): Promise<void> {
  return assertModule().notifyAuthRefreshed();
}

export async function setAuthTokens(tokens: AuthTokens): Promise<void> {
  return assertModule().setAuthTokens(toNativeAuthTokens(tokens));
}

export async function getValidAccessToken(): Promise<string | null> {
  return assertModule().getValidAccessToken();
}

export async function clearAuthTokens(): Promise<void> {
  return assertModule().clearAuthTokens();
}

// ============================================================================
// Maintenance API
// ============================================================================

/**
 * Remove completed sessions older than specified time.
 * @param olderThanMs - Time threshold in milliseconds (default: 0 = all completed)
 * @returns Number of sessions purged
 */
export async function purgeCompletedSessions(
  olderThanMs: number = 0
): Promise<number> {
  if (olderThanMs < 0) {
    olderThanMs = 0;
  }
  return assertModule().purgeCompletedSessions(olderThanMs);
}

/**
 * Clean up temporary files created during upload process.
 * @returns Number of files cleaned
 */
export async function cleanupTempFiles(): Promise<number> {
  return assertModule().cleanupTempFiles();
}

// ============================================================================
// Utility API
// ============================================================================

/**
 * Extract MD5 hashes from local files.
 * @param fileUris - Array of local file URIs
 * @returns Array of MD5 results (hex strings)
 */
export async function extractMD5FromFiles(
  fileUris: string[]
): Promise<Md5HexResult[]> {
  if (!Array.isArray(fileUris) || fileUris.length === 0) {
    return [];
  }
  return assertModule().extractMD5FromFiles(fileUris);
}
